sdk/tables/data-tables/src/secondaryEndpointPolicy.ts

```diff
@@ -0,0 +1,69 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { OperationOptions } from "@azure/core-client";
+import { PipelinePolicy } from "@azure/core-rest-pipeline";
+
+/**
+ * The programmatic identifier of the tablesSecondaryEndpointPolicy.
+ */
+export const tablesSecondaryEndpointPolicyName = "tablesSecondaryEndpointPolicy";
+export const SecondaryLocationHeaderName = "tables-secondary-endpoint";
+const SecondaryLocationAccountSuffix = "-secondary";
+
+/**
+ * Policy that would replace the Primary Endpoint with the secondary endpoint
+ * when the `tables-secondary-endpoint` is set in the request
+ */
+export const tablesSecondaryEndpointPolicy: PipelinePolicy = {
+  name: tablesSecondaryEndpointPolicyName,
+  sendRequest: async (req, next) => {
+    // Only replace the URL if the SecondaryLocationHeader is set
+    if (req.headers.get(SecondaryLocationHeaderName)) {
+      // Since the header is for internal use only, clean it up.
+      req.headers.delete(SecondaryLocationHeaderName);
+      // Calculate and update the secondary url
+      req.url = getSecondaryUrlFromPrimary(req.url);
+    }
+
+    return next(req);
+  },
+};
+
+/**
+ * Utilility function that injects the SecondaryEndpointHeader into an operation options
+ */
+export function injectSecondaryEndpointHeader(options: OperationOptions): OperationOptions {
+  const requestOptions = options.requestOptions;
+  const headerToInject = { [SecondaryLocationHeaderName]: "true" };
+  if (!requestOptions) {
+    return {
+      ...options,
+      requestOptions: { customHeaders: headerToInject },
+    };
+  }
+
+  if (!requestOptions.customHeaders) {
+    return {
+      ...options,
+      requestOptions: { ...requestOptions, customHeaders: headerToInject },
+    };
+  } else {
+    requestOptions.customHeaders = { ...requestOptions.customHeaders, ...headerToInject };
+    return { ...options, requestOptions };
+  }
```

Comment:
style nit: can simplify this a bit

```suggestion
  const headerToInject = { [SecondaryLocationHeaderName]: "true" };

  return {
    ...options,
    requestOptions: {
      ...options.requestOptions,
       customHeaders: {
         ...options.requestOptions?.customHeaders,
         ...headerToInject
       }
    }
  };
```